static-analysis/teammates-macker.xml

```diff
@@ -1,45 +1,340 @@
+                            <include class="teammates.ui.template.StudentProfileEditBox" />
+                        </from>
+                    </allow>          
+                </deny>
+            </access-rule>
+        </ruleset>
+        <ruleset name="UiComponent: Controller should be self-contained">
+            <access-rule>
+                <deny>
+                    <from pattern="productionCode" />
+                    <to class="teammates.ui.controller.**" />
+                    <allow>
+                        <from class="teammates.ui.controller.**" />
+                    </allow>          
+                </deny>
+            </access-rule>
+        </ruleset>
+        
+        <ruleset name="LogicComponent: Each logic can only access its corresponding DB (e.g. AccountsLogic -> AccountsDb)">
+            <foreach var="logicModule" class="teammates.logic.core.(*)Logic">
+                <pattern name="DbModule" class="teammates.storage.api.${logicModule}Db" />
+                
+                <access-rule>  
+                    <message>${logicModule} can only access its corresponding DB</message>
+                    <deny> 
+                        <from class="teammates.logic.core.${logicModule}Logic" />
+                        <to class="teammates.storage.api.**" /> 
+                        <allow>
+                            <to pattern="DbModule" />
+                        </allow>
+                    </deny>
+                </access-rule>
+            </foreach>
+        </ruleset>
+        <ruleset name="LogicComponent: BackDoorServlet should not access logic:core">
+            <access-rule>
+                <deny>
+                    <from class="teammates.logic.backdoor.BackDoorServlet" />
+                    <to class="teammates.logic.core.**" />      
+                </deny>
+            </access-rule>
+        </ruleset>
+        
+        <ruleset name="StorageComponent: storage::entity should only be accessed by storage::api and common:datatransfer">
+           <access-rule>
+                <deny>
+                    <from pattern="productionCode" />
+                    <to class="teammates.storage.entity.**" />
+                </deny>
                 <allow>
-                    <!-- allow dependency from inner class -->
-                    <from class="${testCasesPackage}.**$**" />
+                    <from>
+                        <include class="teammates.storage.entity.**" />
+                        <include class="teammates.storage.api.**" />
+                        <include class="teammates.common.datatransfer.attributes.**" />
+                    </from>
                 </allow>
                 <allow>
-                    <!-- allow dependency to inner class -->
-                    <to class="${testCasesPackage}.**$**" />
+                    <!-- Violation that needs fixing -->
+                    <from class="teammates.logic.core.FeedbackResponsesLogic" />
```

Comment:
To catch this violation, you can a rule similar to `"UiComponent: Only *Action can touch Logic API`: `LogicComponent: Only access StorageComponent via Storage API`.